### PR TITLE
Add BOOL_FALSE guard to optimize empty container case

### PR DESCRIPTION
**Pull request by wconstab**
There is a fast way to implement a guard for an empty dict, which is to check its bool() value.

However, we can't use this guard in general, since we can only safely apply it at runtime if the runtime value actually is a dict (or, another type that works with 'bool' in the same way).  A counterexample is when a tensor is passed instead of a dict, and throws on bool() operator.

So we can put a type check in the guard, but that is slow enough it defeats the purpose.

Instead, we note that for the case of NNModuleVariables (which are specialized NNModules not unspecialized ones), we already have a hook in place to invalidate the guards if setattr is called.  I am claiming that setattr is the only way that the type of a property on an NNModule could change.  If I'm right, then it's safe to (a) only use this guard for NNModuleVariables, (b) not do a type check inside the guard.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94917
* #94951
* #92125
* __->__ #95248

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/guards.py [hunk 1]
```diff
@@ -169,6 +169,26 @@ def TYPE_MATCH(self, guard: Guard):
         code = f"___check_type_id({self.arg_ref(guard)}, {obj_id})"
         self._produce_guard_code(guard, [code])
 
+    def BOOL_FALSE(self, guard: Guard):
+        # Guard on the runtime value being 'False',
+        # can be faster than seemingly equivalent checks like DICT_KEYS for empty dict
+        ref = self.arg_ref(guard)
+        value = self.get(guard.name)
+        t = type(value)
```
Comment (voznesenskym) [inline torch/_dynamo/guards.py:177]: t unused?

File: torch/_dynamo/guards.py [hunk 2]
```diff
@@ -169,6 +169,26 @@ def TYPE_MATCH(self, guard: Guard):
         code = f"___check_type_id({self.arg_ref(guard)}, {obj_id})"
         self._produce_guard_code(guard, [code])
 
+    def BOOL_FALSE(self, guard: Guard):
+        # Guard on the runtime value being 'False',
+        # can be faster than seemingly equivalent checks like DICT_KEYS for empty dict
+        ref = self.arg_ref(guard)
+        value = self.get(guard.name)
```
Comment (voznesenskym) [inline torch/_dynamo/guards.py:176]: value unused?

File: torch/_dynamo/variables/builder.py [hunk 3]
```diff
@@ -291,7 +291,17 @@ def _wrap(self, value):
...
         ):
-            guards = self.make_guards(GuardBuilder.DICT_KEYS)
+            if not value and self.get_source().is_nn_module():
+                # It is faster to guard on 'false' property than to guard
+                # on actual dict keys, but we can't do this fast guard in general because
+                # it omits a crucial type check that ensures the value is actually still a dict at runtime.
+
+                # Why is this OK for (specialized) nnmodules? We set up a setattr hook
+                # to check for mutations, and I claim this would intercept any attempt to change
+                # the type of a dict on the  module.
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builder.py:301]: Does this matter? Let's imagine a few cases:

1) `is not value` -> BOOL_FALSE. Then, someone populates the dict. Guard correctly fails `not` check. Looks good.
2) `is not value` -> BOOL_FALSE. Then, someone swaps the type on the dict, but it is still `not foo == True` ... so it is now a NoneType, and your hook would catch it. 

However, another way would be to just have BOOL_FALSE also call `self.TYPE_MATCH(guard)` in the BOOL_FALSE, and then you (1) know that BOOL_FALSE is always protected, without having it to rely on outside logic only applicable to dicts on modules (2) is more idiomatic with how we compose guards.

Comment (wconstab) [reply torch/_dynamo/variables/builder.py:301]: TYPE_MATCH is slow. that's my problem.  (I tried this first.  Then I tried taking simple / minimal pieces out of type check and using them here.  i could not find a fast enough way to do a type check)

File: torch/_dynamo/guards.py (hunk 1 above)
Comment (wconstab) [reply torch/_dynamo/guards.py:177]: yea, ill clean these up if i land it.  I was using them in the commented out versions of the code that had type checks.

File: torch/_dynamo/variables/builder.py (hunk 3 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:301]: Hmm, I thought ODICT_KEYS/DICT_KEYS  was the problem?

Comment (wconstab) [reply torch/_dynamo/variables/builder.py:301]: well it was also slow. but i found type check is also slow

Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:301]: @jansel thoughts?

**Review by voznesenskym — APPROVED** [review]

**Pull request closed without merging**